torch/_dynamo/variables/torch_function.py

```diff
@@ -0,0 +1,195 @@
+    """
+
+    @staticmethod
+    def create(
+        tx,
+        tensor_variable,
+        torch_function_fn,
+        subclass_type,
+        **kwargs,
+    ):
+        var = TensorWithTFOverrideVariable(
+            tensor_variable,
+            torch_function_fn,
+            subclass_type,
+            **kwargs,
+        )
+        # stash the subclass type to rewrap an output tensor if needed
+        if var.global_class_name() not in tx.output.global_scope:
+            tx.output.install_global(var.global_class_name(), subclass_type)
+
+        return var
+
+    def __init__(
+        self,
+        tensor_variable,
+        torch_function_fn,
+        subclass_type,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.tensor_variable = tensor_variable
+        self.torch_function_fn = torch_function_fn
+        self.subclass_type = subclass_type
+
+    def as_proxy(self):
+        return self.tensor_variable.as_proxy()
+
+    def python_type(self):
+        return self.subclass_type
+
+    def torch_function_var(self, tx):
+        from .builder import VariableBuilder
+
+        source = AttrSource(
+            AttrSource(self.source, "__torch_function__"),
+            "__func__",
+        )
+        return VariableBuilder(tx, source)(self.torch_function_fn)
+
+    def subclass_type_var(self):
+        return UserDefinedClassVariable(self.subclass_type)
+
+    def global_class_name(self):
+        return f"__subclass_{self.subclass_type.__name__}"
+
+    def call_torch_function(self, tx, fn, types, args, kwargs):
+        return call_torch_function(
+            tx,
+            self.subclass_type_var(),
+            build_torch_function_var(tx, self.torch_function_fn, self.source),
+            fn,
+            types,
+            args,
+            kwargs,
+        )
+
+    def call_method(
+        self,
+        tx,
+        name,
+        args: "List[VariableTracker]",
+        kwargs: "Dict[str, VariableTracker]",
+    ) -> "VariableTracker":
+        # This code block implements inlining the __torch_function__ override
+        # of `call_method`.
+        if tx.output.torch_function_enabled:
+            import torch
+            from .builder import SourcelessBuilder
+
+            options = VariableTracker.propagate(self, args, kwargs.values())
+            args = [self] + list(args)
+            # [Note: __torch_function__] Currently we only support methods that are defined on tensor
+            # we will graph break in other cases this will need a bigger overhaul of extracting methods/comparing them for equality
+            func_var = SourcelessBuilder()(tx, getattr(torch.Tensor, name))
+            all_args = args + tree_flatten(kwargs)[0]
+            types = TupleVariable(
+                [
+                    a.subclass_type_var()
+                    for a in all_args
+                    if isinstance(a, TensorWithTFOverrideVariable)
+                ]
+            )
```

Comment:
Why isn't this using `get_overloaded_args`?

Reply:
So I meant to check with you, even on a method call, the `self` tensor might not be the prioritized `__torch_function__` impl? I assumed that we would be calling the `self` impl, it seems a little strange to me that you might call a different impl when calling a method no? If those are the correct semantics though I'm happy to do it.

Reply:
I'll also add a test case for this

Reply:
You indeed may not prioritize the self impl on a method call. For example, if you have Foo and FooSubclass, foo.add(foo_subclass) will preferentially dispatch to the subclass first.